### PR TITLE
CMake: Respect BZIP2_DIR in FindBZIP2.cmake

### LINKED ISSUE #5063: FindBZIP2.cmake should respect BZIP2_DIR variable.
I recently had a problem where our config ignored `DBZIP2_DIR=` and picked up `bzip2` from `/usr/*`:
```
#        DEAL_II_WITH_BZIP2 set up with external dependencies
#            BZIP2_VERSION = 1.0.6
#            BZIP2_DIR = /spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/bzip2-1.0.6-4vcfqu3netbuizhbtflvsrliinx7jogd
#            BZIP2_INCLUDE_DIRS = /usr/include
#            BZIP2_LIBRARIES = /usr/lib64/libbz2.so
```
Apart from the unwanted surprise, the problem is that `/usr/lib64/libbz2.so` is there on log-in nodes, but isn't on calculation nodes.

My understanding that for this case we eventually call https://github.com/dealii/dealii/blob/master/cmake/macros/macro_configure_feature.cmake#L126

Judging from [CMake docu](https://cmake.org/cmake/help/v3.0/command/find_package.html), i think we should add `NO_CMAKE_PATH` so that it starts from point (2):

> \<package\>_DIR
CMAKE_PREFIX_PATH
CMAKE_FRAMEWORK_PATH
CMAKE_APPBUNDLE_PATH

@tamiko ping.

p.s. I know that I can override it with `-DBZIP2_FOUND=TRUE,...`

### PR DESCRIPTION
**Pull request by tamiko**
Closes #5063

### CODE REVIEW

**Comment by tamiko** [comment]
@davydden Please test this patch.

**Comment by davydden** [comment]
i will, but a but tomorrow

**Comment by davydden** [comment]
@tamiko see https://github.com/dealii/dealii/issues/5063#issuecomment-328900051

File: cmake/modules/FindBZIP2.cmake [hunk 1]
```diff
@@ -23,15 +23,19 @@
...
-#
+SET(BZIP2_DIR "" CACHE PATH "An optional hint to a BZIP2 installation")
+SET_IF_EMPTY(BZIP2_DIR "$ENV{BZIP2_DIR}")
+
+SET(_cmake_prefix_path_backup "${CMAKE_PREFIX_PATH}")
 
 # temporarily disable ${CMAKE_SOURCE_DIR}/cmake/modules for module lookup
 LIST(REMOVE_ITEM CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake/modules/)
+
+SET(CMAKE_PREFIX_PATH ${BZIP2_DIR} ${_cmake_prefix_path_backup})
```
Comment (davydden) [inline cmake/modules/FindBZIP2.cmake]: let's instead add `NO_CMAKE_PATH` to `find_package` and thus ignore

> CMAKE_PREFIX_PATH
CMAKE_FRAMEWORK_PATH
CMAKE_APPBUNDLE_PATH

and go directly to

> <package>_DIR
CMAKE_PREFIX_PATH
CMAKE_FRAMEWORK_PATH
CMAKE_APPBUNDLE_PATH

which solves this issue for all packages.

**Comment by davydden** [comment]
unless there are further suggestions, feel free to merge in a day or so.

**Comment by tjhei** [comment]
With this patch we will look at ``BZIP2_DIR`` first, but continue to check the default search paths after that, correct? Is this what we want? I think I would prefer not finding the library instead.

**Comment by davydden** [comment]
> Is this what we want?

yes

> I think I would prefer not finding the library instead.

why? If I point to `bzip2` via the exact prefix, I expect the build system to actually find it there.

**Comment by tjhei** [comment]
> why?

Because it will silently and unexpectedly use a (possibly broken) system install of a library.

**Comment by davydden** [comment]
> Because it will silently and unexpectedly use a (possibly broken) system install of a library.

oh, you mean when `BZIP2` directly is rubbish and does not point to the library? But that's exactly the current behaviour -- it will find whatever wherever.

**Comment by tamiko** [comment]
@tjhei The current behavior is to use a number of hints and default locations in order to find an external feature. This is true for all other 21 external dependencies we have.

**Comment by tjhei** [comment]
> @tjhei The current behavior is to use a number of hints and default locations in order to find an external feature. This is true for all other 21 external dependencies we have.

Yes, I am aware of that. Maybe I shouldn't have hijacked this PR for the discussion.

**Comment by tamiko** [comment]
@tjhei Well then, hit the merge button :-P

But just to iterate on your suggestion. I would very much prefer not to change such behavior. Simply because this creates much more confusion than benefit. Further, overriding the default behavior of CMake will significantly increase complexity. (And it is actually not clear to me how to reliably "prevent" cmake modules from still finding external dependencies.)

**Pull request merged**